### ISSUE TITLE
Add sorting ability to list of entries

### ISSUE THREAD

**Issue by phae** [netlify/netlify-cms#54]
User needs:
- to more efficiently be able to locate a post when browsing manually

So, we should add some basic filters to reorder the list of posts that have already gone live.  

Minimum options for now.

Acceptance criteria:
- [ ] User can click a link to re-order the posts from newest to oldest, and visa versa
- [ ] User can click a link to re-order the posts in alphabetical and reverse-alphabetical
#50 is prerequisite for this.

**Comment by loremipson** [comment]
Requesting some more flexibility here. What do you all think about accepting some optional yaml in the config that would result in both additional fields being displayed in the entry list, as well as making them sortable?

Something like: 

```yaml
collections:
  - name: posts
     label: Posts
     folder: src/posts
     create: true
     fields:
       - { name: title, label: Title, widget: string }
       - { name: publishedDate, label: Published Date, widget: datetime }
       - { name: editedDate, label: Edited Date, widget: datetime }
       - { name: body, label: Body, widget: markdown }
     sortableFields: ["title", "publishedDate", "editedDate"]
```

**Comment by mittalyashu** [comment]
I have created this. It is similar to what @phae said.

<img width="393" alt="group 4" src="https://user-images.githubusercontent.com/29014463/45480148-ec5e8f00-b765-11e8-8b6a-9e6f05d9f5ca.png">

Check the [source file](https://www.figma.com/file/0bTxEnrHCNM32hDPDQJvrzbr/NetlifyCMS-Design?node-id=86%3A154).

**Comment by tomrutgers** [comment]
Looks good! I would split the ascending / descending functionality from the select box though. So a select box to select the sort modifier and one button to switch between ascending and descending.

**Comment by mittalyashu** [comment]
That will take too much of space from design prospective. Why not just keep inside a single component.

I know the title of the button is wrong, but the image is just to show the design.

**Comment by enriquemorenotent** [comment]
Is this request really 2 years old? This is so important!

**Comment by Herohtar** [comment]
@mittalyashu Taking more space is preferable to having a confusing UI component. "Ascending" and "Descending" are sort directions and should be separate from the list of fields to sort by. Not to mention that the way there displayed there implies you can somehow select both at once.

**Comment by tomrutgers** [comment]
I agree with @Herohtar. Also: it does not have to take up extra space. If you move the asc/desc icon out of the select box and use it as a button you'd have the outcome I mean.

**Comment by mittalyashu** [comment]
Okay, I will try and do some brainstorming.

**Comment by JasonHanley** [comment]
Very impressed with many aspects of this project, but from the perspective of someone evaluating different static CMS tools, having content show up in hard-coded reverse-alphabetical order is a fairly glaring deficiency, and makes it unsuitable for production use. Not being a React expert, I can't even find where in the source to change this.

I'd be happy just to have a simple config option control this, and that might be simpler to implement.
`sort: [fieldname] [asc|desc]`

In the meantime, is someone able to point me to the source where this is hard-coded?

**Comment by enriquemorenotent** [comment]
@JasonHanley That is an interesting thought. I never considered hacking it myself :D

I have not really taken time to analize it well, but my bet would be in this file:

https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-core/src/components/Collection/Entries/EntryListing.js

Either line: #50 or #55. Not sure which one of those 2 methods are used, but I would bet that one of those 2 arrays (or maybe boths) need a `sort` function.

**Comment by erquhart** [comment]
Note that this is more complex than it seems - in many cases you're dealing with a paginated list, in which case more entries load as you scroll. Sorting would therefore have to be supported by the backend in use, or else we need to load all pages up front before sorting.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by erquhart** [comment]
I spoke w/ @Benaiah recently about using sha's to pseudo cache. The big 3 Git hosts all provide some kind of commit diffing via API:

- [GitHub](https://developer.github.com/v3/repos/commits/#compare-two-commits)
- [GitLab](https://docs.gitlab.com/ee/api/commits.html#get-the-diff-of-a-commit)
- [Bitbucket](https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/diff/%7Bspec%7D)

With the ability to see files that changed between two commits, we can create a sort of index in a data file in your repo, updating changed bits on load, and never needing to download all entries (except to create the original cache, or if the cache is somehow lost).

Still some work to do on the concept, but it does feel feasible.

**Comment by Undistraction** [comment]
It seems to me that currently we have paginated entries, but at the expense of:

- ability to order entries 
- ability to filter entries
- ability to search entriea

Implementation of the above three items is a pretty critical feature for any CMS if it is to scale at all. Given that it seems that it is unlikely that the above three features are going to be implemented across all backends due to differences in how they handle pagination, how about making the pagination of results optional? This would allow us the above features to be implemented for collections that are not set to paginate. I know the feedback from the people using my build is that it is virtually unusable without a means of searching or ordering entries, whereas pagination is just a nice-to-have feature that has very little or no effect for relatively small collections.

**Comment by erquhart** [comment]
We didn't choose pagination - the Bitbucket and GitLab (and pretty much every normal API out there) returns paginated results. Some repos using Netlify CMS have thousands of entries in a single collection, so requiring that they all be loaded up front is a non-starter. As GitLab and Bitbucket support were the most requested features by a wide margin, we had to take the good with the bad.

Again, we can start this off with a backend agnostic solution in core and an implementation in the GitHub backend. I don't love having features supported in only one backend, but it's better than nothing.

**Comment by Undistraction** [comment]
> Some repos using Netlify CMS have thousands of entries in a single collection, so requiring that they all be loaded up front is a non-starter.

I agree, and I appreciate that most apis will paginate results, that's why I'm suggesting making pagination something that can be opted out of. For many, probably the majority of cases, there is little or no need for pagination, and it would be fine to load everything up front. Just suggesting that by allowing a collection to opt out of pagination, it could at the same time opt-in to ordering entries, filtering entries and searching entries.

**Comment by Benaiah** [comment]
Opting out of pagination by loading all posts up-front is possible - there’s existing code to do so for search in the GitLab and BitBucket backends. It could be an option for small sites, though the latency upfront can become large very quickly. The CMS would also have to load them every time it detects a change in the repo, meaning this cost would be incurred multiple times in a single CMS session.

**Comment by erquhart** [comment]
That's where [commit diffing](https://github.com/netlify/netlify-cms/issues/54#issuecomment-432471818) comes in.

**Comment by Jones-S** [comment]
I see that this issue is included in the dev cycle ending on 29th of March. Will this mean that we can soon see some sorting within the netlify cms?

**Comment by barbalex** [comment]
I was almost decided on using netlify-cms for two projects. Then I realized that not being able to sort was a no-go for my client :-(

I understand the problems and realize it is hard to solve. Aside from this: great project!

**Comment by erquhart** [comment]
@barbalex question: we're about to implement this, but initially it will be config driven, so you can set the sort in config but won't be able to change it in the UI. Regardless of whether it's too late to actually use it for your client, I'm curious whether what I'm describing would satisfy your requirements?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by barbalex** [comment]
@erquhart Yes, config would suffice as entries would be sorted by publishing date, descending. Plus: I can wait a few weeks. Looking forward to this feature!

**Comment by otaviowillow** [comment]
@erquhart is this still on the roadmap?

**Comment by erquhart** [comment]
Absolutely, just didn't get it out as soon as I had hoped.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by snoopdouglas** [comment]
Great that this is being developed now. Sorting is the last major hurdle for our clients too.

How is this going to work with sorting by custom fields? Is this now part of the acceptance criteria as per @loremipson's [suggestion](https://github.com/netlify/netlify-cms/issues/54#issuecomment-390244193)?

With hundreds of pages, I'd imagine this could get quite slow as we'd be bombarding <your git host> with requests, so will the CMS maintain in-repo index files? eg. `/.netlify-cms/indexes/${collection}/${field}.yml`

**Comment by erquhart** [comment]
That's a part of the plan, I expect repo indexes to play a big role in many features and improvements in the future. Initially this will likely all be in memory. Perf won't be great for large collections, but it will initially be configuration driven (no UI sort selection), so the impact will be limited.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by wing5wong** [comment]
Waiting on this too - just been asked whether or not this is possible to do rather than the current alphabetical.

config only is fine - specifically wanting reverse date order (newest entries at top of list)

**Comment by nratter** [comment]
Any updates on this? Config only is fine for our case as well.

**Comment by erquhart** [comment]
Still coming. Became a part of a bigger effort (🤦‍♂) but will be out soon.

**Comment by alfchee** [comment]
I would like you may add this feature to the CMS, it would be great to be a able to decide how to sort the items of different collections in the CMS.

**Comment by BIIGitHub** [comment]
Any updates on this??

**Comment by shushugah** [comment]
Bump, my file names are numeric based, and it's annoying that right now it sorts by, without any option to sort by date
```
1.md
10.md 
11.md
2.md
3.md
```

**Comment by marcneander** [comment]
Would love an update on this.

Currently no way to sort in a decent way? Best workaround I've found is using "{{year}}-{{month}}-{{day}}" or similar as slug. However, when updating an item NetlifyCMS is not updating filenames.

Am I doing something wrong or are we missing this very basic and crucial feature?

**Comment by erezrokah** [comment]
Best thing you can do is up vote the issue as we use it for prioritization. This is currently number 4: https://github.com/netlify/netlify-cms/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by bencao** [comment]
definitely an up vote for this feature.

**Comment by cloudcrestconsulting** [comment]
Any update on this?

**Comment by erezrokah** [comment]
This is next on our list. We needed to pay off some tech debt to make this happen (https://github.com/netlify/netlify-cms/pull/3316)

**Comment by JeffML** [comment]
I would hope we could filter by draft/published status as well.

**Comment by erquhart** [comment]
Latest decisions:
- Implement for all backends
- Sort in UI!
- Defaults:
  - "Updated" - uses Git to sort files by most recently changed
  - Inferred fields - title and date
- Configuration can be used to sort by any field

**Comment by erezrokah** [comment]
Example:
![image](https://user-images.githubusercontent.com/26760571/76437531-fa708880-63b9-11ea-8ba7-8d8f3c187c1b.png)

**Comment by chocobuckle** [comment]
@erquhart By "sort in ui", do you mean we'll be able to click and drag items to new positions?

**Comment by erezrokah** [comment]
> I'd be happy just to have a simple config option control this, and that might be simpler to implement.
> `sort: [fieldname] [asc|desc]`

@chocobuckle I think he means that you can control sorting from the UI. Drag and drop is more related to https://github.com/netlify/netlify-cms/issues/475

**Comment by ChrisLasar** [comment]
Just in case you have note considered it yet: A couple of days ago I found a wonderful "a table enhancing plug-in for the jQuery Javascript library" on Github, which might add wonderful sorting,  filtering and grouping features to the great NetlifyCMS solution. See https://github.com/DataTables/DataTables for more information.

**Comment by austincondiff** [comment]
In `v3-ui-redesign` we are working on exactly this. We’ve created this list view where you can click on the table headings and also drag items in a custom order. 

Check it out here... 
https://storybook.netlifycms.org/?path=/story/components-table--table

**Comment by wing5wong** [comment]
Will this implementation deal with a large list as well - say 200 + items

**Comment by austincondiff** [comment]
@wing5wong I plan to virtualize these rows with `react-window`. So it should handle 200+ items easily.

**Comment by didia** [comment]
Wow! Enfin! Congratulations on closing this one. It's been more than 3 years waiting for this! Good Job team!

**Comment by jacobpake** [comment]
Glad to be able to unsubscribe from this, thank you team

**Comment by equivalentideas** [comment]
Congrats on this implementation @erezrokah @erquhart it works really well :bouquet: :)

My team have a long lists of post and want to have the latest post first, because it's not always easy to remember the name of a post you're looking for to find it and edit it. Latest-post-first is how we present posts on the frontend, and it's not obvious to them why it would be different on the back-end.

Is it possible to set a `date` field to be the default from `sortableFields` and control the direction? or would that be a new feature?

**Comment by erezrokah** [comment]
@equivalentideas Thanks 😄
The reason for not to have a default sort was that sorting requires reading all entries (since it's done client side) which can take some time with large collections, and once a user sorts a collection, that preference is persisted so it's a one time thing.

With that said, since this is not the first time I've heard of that request it's looks like it would make sense to be able to opt in to a default sort method with the cost of a slow first load.
Does that makes sense? Do you mind opening a feature request with your suggestion?
